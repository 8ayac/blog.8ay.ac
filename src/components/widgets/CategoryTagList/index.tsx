import { CategoryTag } from '@/src/components/elements/CategoryTag';
import { FlexContainerDiv } from '@/src/shared/styles/abstractStyledComponents';
import styled from '@emotion/styled';
import { faTags } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import React from 'react';

const S = {
  ComponentWrapperDiv: styled(FlexContainerDiv)`
    flex-flow: row wrap;

    > * {
      margin-right: 0.7rem;
      margin-bottom: 0.7rem;
    }
  `,

  PreIconWrapperDiv: styled(FlexContainerDiv)`
    align-items: center;
    color: ${(props) => props.theme.color.text.primaryLight};
  `,
};

export const CategoryTagList: React.FC<{
  tags: string[];
  withPreIcon?: boolean;
}> = ({ tags, withPreIcon = false }) => (
  <>
    <S.ComponentWrapperDiv>
      {withPreIcon && (
        <S.PreIconWrapperDiv>
          <FontAwesomeIcon icon={faTags} />
        </S.PreIconWrapperDiv>
      )}
      {[...new Set(tags)].sort().map((tag) => (
        <CategoryTag key={tag} name={tag} />
      ))}
    </S.ComponentWrapperDiv>
  </>
);
