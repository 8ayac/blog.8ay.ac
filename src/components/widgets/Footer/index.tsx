import { theme } from '@/src/constants/theme';
import styled from '@emotion/styled';
import React from 'react';

const S = {
  Footer: styled.footer`
    padding: 3rem 0;
    font-size: 1.5rem;
    text-align: center;
  `,

  CopyRightParagraph: styled.p``,

  AboutGAParagraph: styled.p`
    color: ${theme.color.text.primaryLight};
  `,
};

export const Footer: React.FC = () => (
  <S.Footer>
    <S.CopyRightParagraph>
      Â© 2021 blog.8ay.ac All Rights Reserved.
    </S.CopyRightParagraph>
    <S.AboutGAParagraph>
      This website uses Google Analytics, a web analytics service by Google.
      Google Analytics uses cookies, which are placed on your computer to help
      the website analyze how you use the site. You can prevent the the
      collection of your data by disabling cookies in your browser.
    </S.AboutGAParagraph>
  </S.Footer>
);
