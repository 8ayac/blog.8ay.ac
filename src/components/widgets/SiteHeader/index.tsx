import { BlogTitle } from '@/src/components/elements/BlogTitle';
import { SocialLinkWithIcon } from '@/src/components/elements/SocialLinkWithIcon';
import { BLOG_SUBTITLE, BLOG_TITLE } from '@/src/constants/site';
import { theme } from '@/src/constants/theme';
import { FlexContainer } from '@/src/shared/styles/abstractStyledComponents';
import styled from '@emotion/styled';
import { faTwitter } from '@fortawesome/free-brands-svg-icons';
import React from 'react';

const S = {
  Header: styled.header`
    position: sticky;
    top: 0;
    z-index: 100;
    height: 3.7rem;
    padding: 0.5rem 1rem;
    background-color: ${theme.color.green.dark2};
    border-bottom: 1px solid ${theme.color.green.light};
  `,

  HeaderInner: styled.div`
    display: flex;
    align-items: center;
    justify-content: space-between;
    height: 100%;
    color: white;
  `,

  TitleWrapper: styled(FlexContainer)`
    align-items: center;
  `,

  NavWrapper: styled(FlexContainer)`
    align-items: center;
  `,
};

export const SiteHeader: React.FC = () => (
  <S.Header>
    <S.HeaderInner>
      <S.TitleWrapper>
        <BlogTitle title={BLOG_TITLE} subtitle={BLOG_SUBTITLE} />
      </S.TitleWrapper>

      <S.NavWrapper>
        <SocialLinkWithIcon
          url="http://twitter.com/8ayac"
          description="8ayac"
          icon={faTwitter}
        />
      </S.NavWrapper>
    </S.HeaderInner>
  </S.Header>
);
