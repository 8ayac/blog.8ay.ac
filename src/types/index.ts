export type ArticleAttribute = {
  title: string;
  publishedAt: Date;
  updatedAt: Date;
  tags?: string[];
};

export type Article = ArticleAttribute & {
  body: string;
};
