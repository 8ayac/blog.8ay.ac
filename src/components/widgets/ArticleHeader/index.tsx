import { DateTime } from '@/src/components/elements/DateTime';
import { theme } from '@/src/constants/theme';
import { FlexContainerDiv } from '@/src/shared/styles/abstractStyledComponents';
import { Article } from '@/src/types';
import styled from '@emotion/styled';
import React from 'react';

const S = {
  WrapperHeader: styled.header`
    border-bottom: 0.25rem solid ${theme.color.border.primaryLight};
    margin-bottom: 3rem;
  `,

  ArticleTitleH1: styled.h1`
    font-size: 2.16em;
    margin: 0 0 1rem;
  `,

  DateTimeWrapperDiv: styled(FlexContainerDiv)`
    align-items: center;
    flex-flow: row wrap;

    > * {
      margin-bottom: 0.3rem;
      margin-right: 1rem;
    }
  `,
};

export const ArticleHeader: React.FC<{ article: Article }> = ({ article }) => (
  <S.WrapperHeader>
    <S.DateTimeWrapperDiv>
      <DateTime date={article.publishedAt} description="Published At" />
      <DateTime date={article.updatedAt} description="Updated At" />
    </S.DateTimeWrapperDiv>
    <S.ArticleTitleH1>{article.title}</S.ArticleTitleH1>
  </S.WrapperHeader>
);
