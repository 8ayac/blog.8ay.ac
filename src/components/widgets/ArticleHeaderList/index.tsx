import { ArticleHeader } from '@/src/components/widgets/ArticleHeader';
import { Article } from '@/src/types';

import styled from '@emotion/styled';
import React from 'react';

const S = {
  ComponentWrapperDiv: styled.div`
    display: grid;
    grid-gap: 5%;
    margin: 0 auto;
  `,

  ArticleHeaderWrapperDiv: styled.div`
    h1 {
      font-size: 2.3rem;
    }
  `,
};

export const ArticleHeaderList: React.FC<{ articles: Article[] }> = ({
  articles,
}) => (
  <>
    <S.ComponentWrapperDiv>
      <S.ArticleHeaderWrapperDiv>
        {articles
          .sort(
            (a, b) =>
              new Date(a.publishedAt).getTime() -
              new Date(b.publishedAt).getTime()
          )
          .reverse()
          .map((article) => (
            <ArticleHeader key={article.attributes.id} article={article} />
          ))}
      </S.ArticleHeaderWrapperDiv>
    </S.ComponentWrapperDiv>
  </>
);
