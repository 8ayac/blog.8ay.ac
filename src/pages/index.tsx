import { BodyHeader } from '@/src/components/elements/BodyHeader';
import { ArticleHeaderList } from '@/src/components/widgets/ArticleHeaderList';
import { CategoryTagList } from '@/src/components/widgets/CategoryTagList';
import jsonArticles from '@/src/shared/.content/articles.json';
import { getNumberOfArticlesWithATag } from '@/src/shared/utils';
import { Article } from '@/src/types';
import styled, { StyledComponent } from '@emotion/styled';
import { GetStaticProps, InferGetStaticPropsType, NextPage } from 'next';
import React from 'react';

const S: { [key: string]: StyledComponent<any> } = {
  Section: styled.section`
    margin-bottom: 4rem;
  `,
};
S.ArticleListSection = styled(S.Section)``;
S.CategoryListSection = styled(S.Section)``;

export const getStaticProps: GetStaticProps = async () => {
  return {
    props: {
      articles: jsonArticles,
    },
  };
};

const TopPage: NextPage<InferGetStaticPropsType<typeof getStaticProps>> = ({
  articles,
}) => {
  return (
    <>
      <S.ArticleListSection>
        <BodyHeader title="Articles" />
        <ArticleHeaderList articles={articles} />
      </S.ArticleListSection>

      <S.CategoryListSection>
        <BodyHeader title="Tags" />
        <CategoryTagList
          tags={(articles as Article[])
            .map((article) => article.tags)
            .flat()
            .map(
              (tag) => `${tag} (${getNumberOfArticlesWithATag(tag, articles)})`
            )}
        />
      </S.CategoryListSection>
    </>
  );
};

export default TopPage;
