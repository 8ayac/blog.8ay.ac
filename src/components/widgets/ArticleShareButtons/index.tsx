import { config } from '@/site.config';
import { Facebook } from '@/src/components/widgets/ArticleShareButtons/CustomShareButton/Facebook';
import { Hatena } from '@/src/components/widgets/ArticleShareButtons/CustomShareButton/Hatena';
import { Line } from '@/src/components/widgets/ArticleShareButtons/CustomShareButton/Line';
import { Pocket } from '@/src/components/widgets/ArticleShareButtons/CustomShareButton/Pocket';
import { Reddit } from '@/src/components/widgets/ArticleShareButtons/CustomShareButton/Reddit';
import { Twitter } from '@/src/components/widgets/ArticleShareButtons/CustomShareButton/Twitter';
import { getArticlePagePath } from '@/src/shared/utils';
import { Article } from '@/src/types';
import styled from '@emotion/styled';
import React from 'react';
import urljoin from 'url-join';

const S = {
  ComponentWrapperSection: styled.section``,

  ButtonListWrapperFlexDiv: styled.div`
    display: flex;
  `,

  ButtonWrapperDiv: styled.div`
    margin-right: 1rem;
    margin-bottom: 0.5rem;
    filter: drop-shadow(
      2px 2px 1px
        ${(props) =>
          props.theme.color.articleShareButtons.buttonWrapperDiv.shadow}
    );

    :hover {
      filter: brightness(120%);
    }
  `,
};

export const ArticleShareButtons: React.FC<{
  article: Article;
  size?: number;
}> = ({ article, size = 32 }) => {
  const shareUrl = urljoin(config.site.rootUrl, getArticlePagePath(article));
  const shareTitle = article.attributes.title;
  const shareButtons = (buttonSize: number) => [
    <Facebook key="facebook" shareUrl={shareUrl} size={buttonSize} />,
    <Twitter
      key="twitter"
      shareUrl={shareUrl}
      shareTitle={shareTitle}
      size={buttonSize}
    />,
    <Hatena
      key="hatena"
      shareUrl={shareUrl}
      shareTitle={shareTitle}
      size={buttonSize}
    />,
    <Line
      key="line"
      shareUrl={shareUrl}
      shareTitle={shareTitle}
      size={buttonSize}
    />,
    <Reddit
      key="reddit"
      shareUrl={shareUrl}
      shareTitle={shareTitle}
      size={buttonSize}
    />,
    <Pocket
      key="pocket"
      shareUrl={shareUrl}
      shareTitle={shareTitle}
      size={buttonSize}
    />,
  ];

  return (
    <>
      <S.ComponentWrapperSection>
        <S.ButtonListWrapperFlexDiv>
          {shareButtons(size).map((shareButton) => (
            <S.ButtonWrapperDiv key={shareButton.key}>
              {shareButton}
            </S.ButtonWrapperDiv>
          ))}
        </S.ButtonListWrapperFlexDiv>
      </S.ComponentWrapperSection>
    </>
  );
};
