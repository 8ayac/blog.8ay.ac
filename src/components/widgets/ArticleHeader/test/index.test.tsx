import { ArticleHeader } from '@/src/components/widgets/ArticleHeader';
import { Article } from '@/src/types';
import { mount } from 'enzyme';
import React from 'react';

describe('ArticleHeader', () => {
  const testArticleData: Article = {
    id: 'test-article',
    title: 'Test Title',
    publishedAt: '2000-01-01 00:00:00 +0900',
    updatedAt: '2000-02-02 00:00:00 +0900',
    tags: ['test-tag1', 'test-tag2', 'test-tag3'],
    body: `Test Body`,
  };

  describe('is rendered correctly', () => {
    test(' to match the snapshot', () => {
      const wrapper = mount(<ArticleHeader article={testArticleData} />);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('has proper style rules', () => {
    const wrapper = mount(<ArticleHeader article={testArticleData} />);

    test('in DateTimeWrapperDiv', () => {
      expect(wrapper.find('DateTimeWrapperDiv')).toHaveStyleRule(
        'align-items',
        'center'
      );
      expect(wrapper.find('DateTimeWrapperDiv')).toHaveStyleRule(
        'flex-flow',
        'row wrap'
      );

      expect(wrapper.find('DateTimeWrapperDiv')).toHaveStyleRule(
        'margin-right',
        '1rem',
        {
          target: '*',
        }
      );
    });

    test('in CategoryListWrapperDiv', () => {
      expect(wrapper.find('CategoryListWrapperDiv')).toHaveStyleRule(
        'margin-top',
        '2rem'
      );
    });
  });
});
