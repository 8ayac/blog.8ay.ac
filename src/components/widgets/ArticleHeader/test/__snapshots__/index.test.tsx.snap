// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleHeader is rendered correctly  to match the snapshot 1`] = `
<WrapperHeader>
  <DateTimeWrapperDiv>
    <DateTime
      date={1999-12-31T15:00:00.000Z}
      description="Published At"
    />
    <DateTime
      date={2000-02-01T15:00:00.000Z}
      description="Updated At"
    />
  </DateTimeWrapperDiv>
  <ArticleTitleH1>
    Test Title
  </ArticleTitleH1>
</WrapperHeader>
`;

exports[`ArticleHeader is rendered correctly to match the snapshot 1`] = `
<WrapperHeader>
  <DateTimeWrapperDiv>
    <DateTime
      date={1999-12-31T15:00:00.000Z}
      description="Published At"
    />
    <DateTime
      date={2000-02-01T15:00:00.000Z}
      description="Updated At"
    />
  </DateTimeWrapperDiv>
  <ArticleTitleH1>
    Test Title
  </ArticleTitleH1>
</WrapperHeader>
`;
